doc/manual/typing.md

```diff
@@ -0,0 +1,559 @@
+see how typed and untyped code interact.
+
+## Interaction between typed and untyped code
+
+### Using typed code inside untyped code
+
+I invite to you come back to our first love, the `filter` example. In the
+previous section, we included everything -- both the definition of `filter` and
+the call site -- in a typed block. More realistically, `filter` would be a typed
+library function (it is actually part of the stdlib as `lists.filter`), but is
+likely to be called from untyped configuration code. Without an additional
+mechanism, this would just ensure that the implementation of `filter` is without
+flaws, but since the call site still escapes the typechecker, we may end up with
+the very same initial problem. Remember, the caller passes an unchecked, invalid
+value that is raising an error from within `filter`.
+
+Hopefully, Nickel prevents us from this happening. Let us see by ourselves:
+
+```nickel
+lists.filter (fun x => if x % 2 == 0 then x else null) [1,2,3,4,5,6]
+```
+
+Result:
+```
+error: Blame error: contract broken by the caller.
+  ┌─ :1:17
+  │
+1 │ forall a. (a -> Bool) -> List a -> List a
+  │                 ---- expected return type of a function provided by the caller
+  │
+  ┌─ repl-input-45:1:67
+  │
+1 │ lists.filter (fun x => if x % 2 == 0 then x else null) [1,2,3,4,5,6]
+  │                                                                   - evaluated to this expression
+  │
+  = This error may happen in the following situation:
+    1. A function `f` is bound by a contract: e.g. `(Num -> Num) -> Num`.
+    2. `f` takes another function `g` as an argument: e.g. `f = fun g => g 0`.
+    3. `f` is called by with an argument `g` that does not respect the contract: e.g. `f (fun x => false)`.
+  = Either change the contract accordingly, or call `f` with a function that returns a value of the right type.
+  = Note: this is an illustrative example. The actual error may involve deeper nested functions calls.
+
+note:
+    ┌─ <stdlib/lists>:160:14
+    │
+160 │     filter : forall a. (a -> Bool) -> List a -> List a
+    │              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ bound here
+
+[...]
+note:
+  ┌─ repl-input-45:1:1
+  │
+1 │ lists.filter (fun x => if x % 2 == 0 then x else null) [1,2,3,4,5,6]
+  │ -------------------------------------------------------------------- (3) calling <func>
+```
+
+We call `filter` from an untyped location, but still get a spot-on error.
+What's is happening? The interpreter protects typed block by a
+[contract](./contracts.md), which is a runtime check. Please refer to the
+dedicated documentation for more details. For now, you can just remember that
+*any type annotation*, wherever it is, gives rise at runtime to a corresponding
+contract check. In other words, `foo: T` and `foo | T` (here `|` is contract
+application, not the row tail separator) behave exactly the same at *runtime*.
+This approach of protecting typed code from untyped code is known as *sound
+gradual typing*.
```

Comment:
Well, I guess in practice not downcasting from `Dyn` freely means inserting safe casts instead of those downcasts? What I remember from the introduction of [this paper](https://www.cs.cornell.edu/~fabianm/papers/nomalive-oopsla17.pdf) and [this other paper](https://dl.acm.org/doi/10.1145/2837614.2837630) was the idea that in practice, the mean to retain soundness was to insert those runtime checks. But once again, maybe the jargon is not so important for this document, and we can drop it.

Reply:
I removed it, in the end. I don't think it adds useful information at this point.